cache/src/main/java/io/envoyproxy/controlplane/cache/Resources.java

```diff
@@ -1,39 +1,143 @@
 import envoy.api.v2.Cds.Cluster;
+import envoy.api.v2.Cds.Cluster.DiscoveryType;
 import envoy.api.v2.Eds.ClusterLoadAssignment;
 import envoy.api.v2.Lds.Listener;
 import envoy.api.v2.Rds.RouteConfiguration;
+import envoy.api.v2.listener.Listener.Filter;
+import envoy.api.v2.listener.Listener.FilterChain;
+import envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManagerOuterClass.HttpConnectionManager;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class Resources {
 
-  static final String TYPE_URL_PREFIX = "type.googleapis.com/envoy.api.v2.";
+  private static final Logger LOGGER = LoggerFactory.getLogger(Resources.class);
+
+  static final String FILTER_ENVOY_ROUTER = "envoy.router";
+  static final String FILTER_HTTP_CONNECTION_MANAGER = "envoy.http_connection_manager";
+
+  private static final String TYPE_URL_PREFIX = "type.googleapis.com/envoy.api.v2.";
+
+  public static final String CLUSTER_TYPE_URL = TYPE_URL_PREFIX + "Cluster";
+  public static final String ENDPOINT_TYPE_URL = TYPE_URL_PREFIX + "ClusterLoadAssignment";
+  public static final String LISTENER_TYPE_URL = TYPE_URL_PREFIX + "Listener";
+  public static final String ROUTE_TYPE_URL = TYPE_URL_PREFIX + "RouteConfiguration";
+
+  public static final List<String> TYPE_URLS = ImmutableList.of(
+      CLUSTER_TYPE_URL,
+      ENDPOINT_TYPE_URL,
+      LISTENER_TYPE_URL,
+      ROUTE_TYPE_URL);
 
   /**
    * Returns the name of the given resource message.
    *
-   * @param xds the xDS resource message
+   * @param resource the resource message
    */
-  public static String getResourceName(Message xds) {
-    if (xds instanceof Cluster) {
-      return ((Cluster) xds).getName();
+  public static String getResourceName(Message resource) {
+    if (resource instanceof Cluster) {
+      return ((Cluster) resource).getName();
     }
 
-    if (xds instanceof ClusterLoadAssignment) {
-      return ((ClusterLoadAssignment) xds).getClusterName();
+    if (resource instanceof ClusterLoadAssignment) {
+      return ((ClusterLoadAssignment) resource).getClusterName();
     }
 
-    if (xds instanceof Listener) {
-      return ((Listener) xds).getName();
+    if (resource instanceof Listener) {
+      return ((Listener) resource).getName();
     }
 
-    if (xds instanceof RouteConfiguration) {
-      return ((RouteConfiguration) xds).getName();
+    if (resource instanceof RouteConfiguration) {
+      return ((RouteConfiguration) resource).getName();
     }
 
     return "";
   }
 
+  /**
+   * Returns all resource names that are referenced by the given collection of resources.
+   *
+   * @param resources the resource whose dependencies we are calculating
+   */
+  public static Set<String> getResourceReferences(Collection<? extends Message> resources) {
+    final ImmutableSet.Builder<String> refs = ImmutableSet.builder();
+
+    for (Message r : resources) {
```

Comment:
looks like lot of if elses ...probably this could be simplified by 
1. stream and filter `ClusterLoadAssignment` and `RouteConfiguration`
2. separate methods for cluster and listener with some ternary operators.

Reply:
I don't think using streams and filters is really going to make this clearer. The bulk of the space consumed in that initial conditional is for the comment explaining *why* we're filtering those message types.